streaming/stream_session.cc

```diff
@@ -45,10 +45,41 @@ static sstables::offstrategy is_offstrategy_supported(streaming::stream_reason r
     static const std::unordered_set<streaming::stream_reason> operations_supported = {
         streaming::stream_reason::bootstrap,
         streaming::stream_reason::replace,
+        streaming::stream_reason::removenode,
+        streaming::stream_reason::decommission,
+        streaming::stream_reason::repair,
+        streaming::stream_reason::rebuild,
     };
     return sstables::offstrategy(operations_supported.contains(reason));
 }
 
+class offstrategy_trigger {
+    sharded<replica::database>& _db;
+    table_id _id;
+    streaming::plan_id _plan_id;
+    replica::column_family& _cf;
+    std::chrono::steady_clock::time_point last_update;
+
+public:
+    offstrategy_trigger(sharded<replica::database>& db, table_id id, streaming::plan_id plan_id)
+        : _db(db)
+        , _id(id)
+        , _plan_id(plan_id)
+        , _cf(_db.local().find_column_family(_id))
+    {
+        _cf.enable_off_strategy_trigger();
+    }
+    void update() {
+        auto now = std::chrono::steady_clock::now();
+        if (now - last_update > std::chrono::seconds(60)) {
```

Comment:
what does this have to do with streaming?
if updating the off strategy trigger needs to be done less frequently, this should be done there, not here.

Reply:
Receiving mutation fragments from streaming triggers the update. The calling side (streaming layer) knows better if it should update less frequently or not.

Here, the streaming code uses the time to decide. In theory, it could use something else, e.g., number of mutation received.

Reply:
But `update_off_strategy_trigger` is cheap, just delaying offstrategy in 5 minutes.
It seems cheaper to do that instead of calling now() yet again and applying this logic to call it at most every minute.

Reply:
update_off_strategy_trigger() will call timer::rearm which calls cancel and arm, add_timer and del_timer etc. Calling a now is cheaper than rearm(). Calling now() again in update_off_strategy_trigger() is not a problem because it happens every 1s. 
The update() is called on the hot path. We'd better optimize if possible.

include/seastar/core/timer.hh
```
    void rearm(time_point until, std::optional<duration> period = {}) noexcept {
        if (_armed) {
            cancel();
        }
        arm(until, period);
    }
```

src/core/reactor.cc

```
template <typename Clock> 
inline
void timer<Clock>::arm(time_point until, std::optional<duration> period) noexcept {
    arm_state(until, period);
    engine().add_timer(this);
}   


template <typename Clock>
inline
bool timer<Clock>::cancel() noexcept {
    if (!_armed) {
        return false;
    }
    _armed = false;
    if (_queued) {
        engine().del_timer(this);
        _queued = false;
    }
    return true;
}
```

Reply:
I think that the right optimization is to keep the trigger time state and let the timer go off and if it sees that the trigger time was extended, just rearm itself to that time.  This way we'd rearm every 5 minutes and the frequent updates would just extend the trigger time if the timer is already armed.

Reply:
What you described is basically what rearm does, i.e., calling update_off_strategy_trigger unconditionally ;-). I want to avoid the reaming all the time for each and every of the mutation we received.

Reply:
Not quite. My suggestion makes rearm cheap so it can be called as frequently as you want to extend the timer expiration,
since it will be re-queued only on the next expiration.
```diff
diff --git a/include/seastar/core/timer.hh b/include/seastar/core/timer.hh
index 39bf0c27..6cd6d7bd 100644
--- a/include/seastar/core/timer.hh
+++ b/include/seastar/core/timer.hh
@@ -86,10 +86,12 @@ class timer {
     scheduling_group _sg;
     callback_t _callback;
     time_point _expiry;
+    std::optional<time_point> _rearm_expiry;
     std::optional<duration> _period;
     bool _armed = false;
     bool _queued = false;
     bool _expired = false;
+    bool _rearm_required = false;
     void readd_periodic() noexcept;
     void arm_state(time_point until, std::optional<duration> period) noexcept {
         assert(!_armed);
@@ -152,8 +154,9 @@ class timer {
     ///        will automatically rearm itself when it expires, using the period
     ///        to calculate the next expiration time.
     void arm(time_point until, std::optional<duration> period = {}) noexcept;
-    /// Sets the timer expiration time. If the timer was already armed, it is
-    /// canceled first.
+    /// Sets the timer expiration time. If the timer was already armed,
+    /// and the current expiration time is later than the updated one,
+    /// it is canceled first.
     ///
     /// \param until the time when the timer expires
     /// \param period optional automatic rearm duration; if given the timer
@@ -161,6 +164,11 @@ class timer {
     ///        to calculate the next expiration time.
     void rearm(time_point until, std::optional<duration> period = {}) noexcept {
         if (_armed) {
+            if (until > _expiry) {
+                _rearm_expiry = until;
+                _period = *period;
+                return;
+            }
             cancel();
         }
         arm(until, period);
@@ -184,12 +192,17 @@ class timer {
         arm(Clock::now() + delta, {delta});
     }
     /// Sets the timer expiration time, with automatic rearming.
-    /// If the timer was already armed, it is canceled first.
+    /// If the timer was already armed, and the rearmed delta is smaller
+    /// than the existing one, it is canceled first.
     ///
     /// \param delta the time when the timer expires, relative to now. The timer
     ///        will also rearm automatically using the same delta time.
     void rearm_periodic(duration delta) noexcept {
         if (_armed) {
+            if (delta > _period) {
+                _rearm_expiry = Clock::now() + delta;
+                return;
+            }
             cancel();
         }
         arm_periodic(delta);
diff --git a/src/core/reactor.cc b/src/core/reactor.cc
index 2c061ddd..25d738ed 100644
--- a/src/core/reactor.cc
+++ b/src/core/reactor.cc
@@ -1382,6 +1382,11 @@ void reactor::complete_timers(T& timers, E& expired_timers, EnableFunc&& enable_
         expired_timers.pop_front();
         t->_queued = false;
         if (t->_armed) {
+          if (t->_rearm_expiry) {
+            t->_expiry = *t->_rearm_expiry;
+            t->_rearm_expiry.reset();
+            queue_timer(t);
+          } else {
             t->_armed = false;
             if (t->_period) {
                 t->readd_periodic();
@@ -1392,6 +1397,7 @@ void reactor::complete_timers(T& timers, E& expired_timers, EnableFunc&& enable_
             } catch (...) {
                 seastar_logger.error("Timer callback failed: {}", std::current_exception());
             }
+          }
         }
     }
     // complete_timers() can be called from the context of run_tasks()
```

Reply:
@asias why do we need this optimization only for streaming? In repair, we do update the timer when handling the row hash in a rpc verb, which happens at row granularity in my understanding. is it that we missed an optimization for repair too?

I am worried about delaying rearm every 60s, and relying on implementation detail of update_off_strategy_trigger() extending the timer for more than 60s. If we apply any optimization, in my opinion, it should be in replica::table. Or we change update_off_strategy_trigger() interface for caller to specify the time point, if we want to push all that into the caller.

Reply:
Because repair handles a batch of mutation fragments per rpc verb. The rearm overhead is already reduced significantly. 

The whole automatic trigger depends on the implementation detail that we have a timer to trigger off-strategy.  If we take a closer look, there is nothing to worry about the optimization. The worst case is that we shorten the trigger time by 60 seconds which is 60 / 5 * 60 = 20%. (If the last fragment arrives within 60s since last trigger, the last fragment will not extend the trigger time to now + 5m). The gain is that we batch thousands or even millions of trigger (rearm). 
We can do something like:  `if (now - last_update > _cf.get_off_strategy_trigger_time() * 0.1)` to get rid of the hard coded numbers. 

```
+    void update() {
+        auto now = std::chrono::steady_clock::now();
+        if (now - last_update > std::chrono::seconds(60)) {
+            sslog.debug("[Stream #{}] Updated offstrategy trigger for ks={}, table={}, table_id={}",
+                _plan_id, _cf.schema()->ks_name(), _cf.schema()->cf_name(), _id);
+            _cf.update_off_strategy_trigger();
+            last_update = std::chrono::steady_clock::now();
+        }
+    }
```

Reply:
> Not quite. My suggestion makes rearm cheap so it can be called as frequently as you want to extend the timer expiration,
> since it will be re-queued only on the next expiration.

It is good to optimize the efficiency of timer::rearm in general.  However, if we can avoid the rearm in high layer more efficiently, I would prefer to do it there.  For example, in your patch above, we would call queue_timer which in turn calls timer-set.hh::insert. This is not going to be more efficient than what we do in this patch `if (now - last_update > std::chrono::seconds(60)) `.  This simple optimization is going to save millions of billions (n(totoal mutation fragment)) insert and push_back calls. 

```
+          if (t->_rearm_expiry) {
+            t->_expiry = *t->_rearm_expiry;
+            t->_rearm_expiry.reset();
+            queue_timer(t);
+          } else {
```

```
bool reactor::queue_timer(timer<steady_clock_type>* tmr) noexcept {
    return _timers.insert(*tmr);
}  
```

```   
 bool insert(Timer& timer) noexcept
    {
        auto timestamp = get_timestamp(timer);
        auto index = get_index(timestamp);

        _buckets[index].push_back(timer);
        _non_empty_buckets[index] = true;

        if (timestamp < _next) {
            _next = timestamp;
            return true;
        }
        return false;
    }
```

Reply:
@avikivity do you see a need for an optimization like I proposed above in https://github.com/scylladb/scylladb/pull/11848#discussion_r1009270316 for timer rearming?

Reply:
FWIW, I think the optimization for the timer::rearm is orthogonal to the optimization at the caller site of timer::rearm. It is nice to have but not a must. Especially, when the caller site can do better optimization.